### PR TITLE
Adding results directory UI option in vstest task

### PR DESCRIPTION
**Pull request by vagisha-nidhi**
image

### CODE REVIEW

File: Tasks/VsTestV2/inputparser.ts [hunk 1]
```diff
@@ -8,6 +8,7 @@ import { AreaCodes, DistributionTypes } from './constants';
 import * as idc from './inputdatacontract';
 import * as Q from "q";
 import * as isUncPath from 'is-unc-path';
+import { stringify } from 'querystring';
```
Comment (vagisha-nidhi) [inline Tasks/VsTestV2/inputparser.ts]: remove if not needed

File: Tasks/VsTestV2/task.json [hunk 2]
```diff
@@ -145,6 +145,15 @@
             "helpMarkDown": "Folder to search for the test assemblies.",
             "groupName": "testSelection"
         },
+        {
+            "name": "resultsDirectory",
+            "type": "string",
+            "label": "Test results directory",
+            "defaultValue": "",
```
Comment (vagisha-nidhi) [inline Tasks/VsTestV2/task.json]: Should we keep the default value as $(Agent.Temp) ?

Comment (ShreyasRmsft) [reply Tasks/VsTestV2/task.json]: $(Agent.Temp)\TestResults should be the default value

File: Tasks/VsTestV2/inputparser.ts (hunk 1 above)
Comment (ShreyasRmsft) [reply Tasks/VsTestV2/inputparser.ts]: we are not using this anywhere?

Comment (vagisha-nidhi) [reply Tasks/VsTestV2/inputparser.ts]: Nope

File: Tasks/VsTestV2/task.json (hunk 2 above)
Comment (vagisha-nidhi) [reply Tasks/VsTestV2/task.json]: Done. The task will take both absolute and relative values. The relative path will be resolved with reference to system.defaultworkingdirectory

File: Tasks/VsTestV2/Strings/resources.resjson/en-US/resources.resjson [hunk 3]
```diff
@@ -22,6 +22,8 @@
   "loc.input.help.tcmTestRun": "Test run based selection is used when triggering automated test runs from the test hub. This option cannot be used for running tests in the CI/CD pipeline.",
   "loc.input.label.searchFolder": "Search folder",
   "loc.input.help.searchFolder": "Folder to search for the test assemblies.",
+  "loc.input.label.resultsDirectory": "Test results directory",
+  "loc.input.help.resultsDirectory": "Folder to store test results. The vstest task cleans up results directory before test run. $(Agent.TempDirectory) is cleaned after each pipeline run. In case the default value is overriden, the directory will not be automatically cleaned. When results directory is provided in both task and runsettings, the runsettings will be given priority.",
```
Comment (PBoraMSFT) [inline Tasks/VsTestV2/Strings/resources.resjson/en-US/resources.resjson]: Folder to store test results. When this input is not specified, results are stored in $(Agent.TempDirectory) by default, which is cleaned at the end of a pipeline run. 
When the default value is overriden, the specified folder will be cleaned by the task before tests are run.

[the remark this answers is not in the document]
Comment (vagisha-nidhi) [reply Tasks/VsTestV2/Strings/resources.resjson/en-US/resources.resjson]: Making this change in the dtaexecutionhost

File: Tasks/VsTestV2/Strings/resources.resjson/en-US/resources.resjson [hunk 4]
```diff
@@ -22,6 +22,8 @@
   "loc.input.help.tcmTestRun": "Test run based selection is used when triggering automated test runs from the test hub. This option cannot be used for running tests in the CI/CD pipeline.",
   "loc.input.label.searchFolder": "Search folder",
   "loc.input.help.searchFolder": "Folder to search for the test assemblies.",
+  "loc.input.label.resultsFolder": "Test results folder",
+  "loc.input.help.resultsFolder": "Folder to store test results. When this input is not specified, results are stored in $(Agent.TempDirectory) by default, which is cleaned at the end of a pipeline run. When the default value is overridden, the specified folder will be cleaned by the task before tests are run.",
```
Comment (ShreyasRmsft) [inline Tasks/VsTestV2/Strings/resources.resjson/en-US/resources.resjson]: The specified/default results directory will always be cleaned up at the start of the vstest task.

File: Tasks/VsTestV2/inputparser.ts [hunk 5]
```diff
@@ -166,6 +166,14 @@ function getTestReportingSettings(inputDataContract : idc.InputDataContract) : i
     inputDataContract.TestReportingSettings.TestRunTitle = tl.getInput('testRunTitle');
     inputDataContract.TestReportingSettings.TestRunSystem = 'VSTS - vstest';
 
+    const resultsDir = path.resolve(tl.getVariable('System.DefaultWorkingDirectory'), tl.getInput('resultsFolder').trim());
```
Comment (ShreyasRmsft) [inline Tasks/VsTestV2/inputparser.ts]: common.testresults

File: Tasks/VsTestV2/inputparser.ts [hunk 6]
```diff
@@ -166,6 +166,14 @@ function getTestReportingSettings(inputDataContract : idc.InputDataContract) : i
...
+    if (utils.Helper.pathExistsAsDirectory(resultsDir.trim()))
```
Comment (ShreyasRmsft) [inline Tasks/VsTestV2/inputparser.ts]: no check here.

File: Tasks/VsTestV2/inputparser.ts [hunk 7]
```diff
@@ -166,6 +166,14 @@ function getTestReportingSettings(inputDataContract : idc.InputDataContract) : i
...
+    {
+        inputDataContract.TestReportingSettings.TestResultsDirectory =  resultsDir;
+        tl.debug("TestResultsFolder: " + resultsDir);
+        ci.publishEvent({ 'TestResultsFolderUi': resultsDir } );
```
Comment (ShreyasRmsft) [inline Tasks/VsTestV2/inputparser.ts]: temp or more than temp path
common.test or
defaultdir 

push this property of the path as you can't directly push path

File: Tasks/VsTestV2/Strings/resources.resjson/en-US/resources.resjson (hunk 4 above)
Comment (vagisha-nidhi) [reply Tasks/VsTestV2/Strings/resources.resjson/en-US/resources.resjson]: Done

File: Tasks/VsTestV2/inputparser.ts (hunk 6 above)
Comment (vagisha-nidhi) [reply Tasks/VsTestV2/inputparser.ts]: Removed

File: Tasks/VsTestV2/inputparser.ts (hunk 7 above)
Comment (vagisha-nidhi) [reply Tasks/VsTestV2/inputparser.ts]: Added. Is there a better way to add this?

File: Tasks/VsTestV2/inputparser.ts [hunk 8]
```diff
@@ -166,6 +166,31 @@ function getTestReportingSettings(inputDataContract : idc.InputDataContract) : i
     inputDataContract.TestReportingSettings.TestRunTitle = tl.getInput('testRunTitle');
     inputDataContract.TestReportingSettings.TestRunSystem = 'VSTS - vstest';
 
+    const resultsDir = path.resolve(tl.getVariable('Agent.TempDirectory'), tl.getInput('resultsFolder'));
+    inputDataContract.TestReportingSettings.TestResultsDirectory =  resultsDir;
+    tl.debug("TestResultsFolder: " + resultsDir);
+    
+    if (resultsDir.startsWith(tl.getVariable('Agent.TempDirectory') + 'TestResults'))
+    {   
+        ci.publishEvent({ 'TestResultsFolderUi': '$(Agent.TempDirectory)/TestResults' } );
```
Comment (ShreyasRmsft) [inline Tasks/VsTestV2/inputparser.ts]: hey, can you add a customer feature name for this?

Other this event will have the same feature name as all the others and filtering on kusto won't be easy.

File: Tasks/VsTestV2/inputparser.ts [hunk 9]
```diff
@@ -166,6 +166,31 @@ function getTestReportingSettings(inputDataContract : idc.InputDataContract) : i
     inputDataContract.TestReportingSettings.TestRunTitle = tl.getInput('testRunTitle');
     inputDataContract.TestReportingSettings.TestRunSystem = 'VSTS - vstest';
 
+    const resultsDir = path.resolve(tl.getVariable('Agent.TempDirectory'), tl.getInput('resultsFolder'));
+    inputDataContract.TestReportingSettings.TestResultsDirectory =  resultsDir;
+    tl.debug("TestResultsFolder: " + resultsDir);
+    
+    if (resultsDir.startsWith(tl.getVariable('Agent.TempDirectory') + 'TestResults'))
```
Comment (ShreyasRmsft) [inline Tasks/VsTestV2/inputparser.ts]: did you verify if this has a slash at the end?

Comment (ShreyasRmsft) [reply Tasks/VsTestV2/inputparser.ts]: better to do path.join

File: Tasks/VsTestV2/inputparser.ts [hunk 10]
```diff
@@ -166,6 +166,31 @@ function getTestReportingSettings(inputDataContract : idc.InputDataContract) : i
...
+    {   
+        ci.publishEvent({ 'TestResultsFolderUi': '$(Agent.TempDirectory)/TestResults' } );
+    }
+    else if (resultsDir.startsWith(tl.getVariable('Agent.TempDirectory')))
+    {
+        ci.publishEvent({ 'TestResultsFolderUi': '$(Agent.TempDirectory)' } );
```
Comment (ShreyasRmsft) [inline Tasks/VsTestV2/inputparser.ts]: please keep an eye out for formatting, use tslint

File: Tasks/VsTestV2/inputparser.ts [hunk 11]
```diff
@@ -166,6 +166,31 @@ function getTestReportingSettings(inputDataContract : idc.InputDataContract) : i
...
     inputDataContract.TestReportingSettings.TestRunSystem = 'VSTS - vstest';
 
+    const resultsDir = path.resolve(tl.getVariable('Agent.TempDirectory'), tl.getInput('resultsFolder'));
+    inputDataContract.TestReportingSettings.TestResultsDirectory =  resultsDir;
+    tl.debug("TestResultsFolder: " + resultsDir);
+    
+    if (resultsDir.startsWith(tl.getVariable('Agent.TempDirectory') + 'TestResults'))
+    {   
+        ci.publishEvent({ 'TestResultsFolderUi': '$(Agent.TempDirectory)/TestResults' } );
+    }
```
Comment (ShreyasRmsft) [inline Tasks/VsTestV2/inputparser.ts]: you can move this to a private function

**Pull request merged**